experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorAnalysis.cpp

```diff
@@ -0,0 +1,94 @@
+#include "IteratorAnalysis.h"
+#include "iterators/Dialect/Iterators/IR/Iterators.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Operation.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/None.h"
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace mlir {
+namespace iterators {
+
+IteratorAnalysis::IteratorAnalysis(Operation *parentOp) : parentOp(parentOp) {
+  parentOp->walk([&](Operation *op) {
+    llvm::TypeSwitch<Operation *, void>(op).Case<SampleInputOp, ReduceOp>(
+        [&](auto op) { buildIteratorInfo(op); });
+  });
+}
+
+llvm::Optional<IteratorAnalysis::IteratorInfo>
+IteratorAnalysis::getIteratorInfo(Operation *op) const {
+  auto it = opMap.find(op);
+  if (it == opMap.end())
+    return llvm::None;
+  return it->getSecond();
+}
+
+llvm::SmallVector<SymbolRefAttr, 3>
+IteratorAnalysis::createFunctionNames(Operation *op) {
```

Comment:
This and the next function need the member variable `uniqueNumber` in order to produce unique names. I guess that that wasn't clear enough. Maybe because of the name? How about renaming to `findUniqueFunctionNames` and `uniqueFunctionName` (where `unique` in the second name is a verb)?

Reply:
Also note that the collision resolution *has* to be done with member variable -- it is *not* sufficient to look up symbols in the module because collisions can occur with Open/Next/Close functions that are only added in the future based on previously pre-assigned names from this analysis.